### PR TITLE
ref(django): preliminary middleware refactor

### PR DESCRIPTION
**Pull request by joshuarli**
Django 2.0 requires the use of "new style" middleware under MIDDLEWARE, removing support for "old style" MIDDLEWARE_CLASSES. In most scenarios, you can use a django.utils.deprecation.MiddlewareMixin to make an old middleware compatible with MIDDLEWARE and MIDDLEWARE_CLASSES. This PR also converts middlewares to use the mixin if it's applicable, and we can flip the switch to MIDDLEWARE separately which requires getsentry to be updated (done in https://github.com/getsentry/getsentry/pull/5627) as well as django-sudo (done in https://github.com/getsentry/sentry/pull/26602).

Due to the nature of the differences, I had to audit all our middleware and in that process I thought some refactoring and cleaning up wouldn't hurt:

- The healthcheck is now moved to the very top, as it should be. Fun fact, under old middleware, any response is processed by ALL middleware. Once we switch to new middleware, we get the new short circuiting semantics which means if a middleware short-circuits by returning a response, only the middleware that executed up to that point will get to process that response. This is something I had to audit for, to make sure we don't depend on that behavior. In this specific example, healthchecks will be very nice when we switch to MIDDLEWARE.

- Removed DecompressBodyMiddleware. Since relay it's no longer needed. Intro'd in https://github.com/getsentry/sentry/pull/9238 to support compressed crashpad requests. Otherwise, we don't have anything to do with compressed requests. Made sure of that in https://github.com/getsentry/sentry/pull/26153, the metric mostly appears with paths to nonexistent python api store, and a few other nonexistent ones like /envelope/.

- Removed NoIfModifiedSinceMiddleware, used by Cramer to debug something in 2014.

- Removed ServicesUnavailableMiddleware, settings.MAINTENANCE isn't used anywhere.

- SentryLocaleMiddleware and SudoMiddleware inherit middleware that already has the mixin. (In the case of SudoMiddleware, this isn't the case until I land my django-sudo updates, done in https://github.com/getsentry/sentry/pull/26602.)

- Future: Could investigate removal of UserActiveMiddleware. At a glance it isn't clear what last_active (py) / lastActive (ts) is used for.

As for the edge cases... I don't think anything needs to be done to process_view, Django still supports that for class-based middlewares (old middlewares with the mixin). So just need to look if process_exception is expecting an exception in process_request.

process_exception in ResponseCodeMiddleware is fine.

process_exception in RequestTimingMiddleware is fine.

process_exception in AuthenticationMiddleware is fine, process_request doesn't raise
AuthUserPasswordExpired anywhere.

### CODE REVIEW

File: src/sentry/middleware/security.py [hunk 1]
```diff
@@ -1,14 +1,13 @@
...
 
     def process_response(self, request, response):
-        # NOTE: there is no `response.setdefault()`
-        if "X-Frame-Options" not in response:
-            response["X-Frame-Options"] = "deny"
-        if "X-Content-Type-Options" not in response:
-            response["X-Content-Type-Options"] = "nosniff"
-        if "X-XSS-Protection" not in response:
-            response["X-XSS-Protection"] = "1; mode=block"
+        response.setdefault("X-Frame-Options", "deny")
```
Comment (joshuarli) [inline src/sentry/middleware/security.py:10]: Verified with devserver that these are still set.

**Comment by BYK** [comment]
>Removed ServicesUnavailableMiddleware, settings.MAINTENANCE isn't used anywhere.

I'm only wary about this one as I'm not sure if ops ever uses it. Have you asked around?

File: src/sentry/conf/server.py [hunk 2]
```diff
@@ -70,8 +70,6 @@ def env(key, default="", type=None):
 
 ADMIN_ENABLED = DEBUG
 
-MAINTENANCE = False
```
Comment (EvanPurkhiser) [inline src/sentry/conf/server.py:73]: Could this have been used in open source? cc @BYK

Comment (wedamija) [reply src/sentry/conf/server.py:73]: Yeah, I'm wondering if ops would ever set this...

File: src/sentry/middleware/proxy.py [hunk 3]
```diff
@@ -105,33 +30,3 @@ def process_request(self, request):
             real_ip = real_ip.split(",")[0].strip()
             real_ip = self._remove_port_number(real_ip)
             request.META["REMOTE_ADDR"] = real_ip
-
-
-class DecompressBodyMiddleware:
```
Comment (EvanPurkhiser) [inline src/sentry/middleware/proxy.py:110]: We ran this by the ingest team yeah?

**Review by wedamija — COMMENTED** [review]
Looks good to me other than just confirming with ops about maintenance mode

File: src/sentry/middleware/proxy.py (hunk 3 above)
Comment (joshuarli) [reply src/sentry/middleware/proxy.py:110]: Yes. Although, I found one thing that cropped up today for release files...

File: src/sentry/conf/server.py (hunk 2 above)
Comment (BYK) [reply src/sentry/conf/server.py:73]: Nope, not in ~~open source~~ self-hosted @EvanPurkhiser

File: src/sentry/middleware/proxy.py (hunk 3 above)
Comment (joshuarli) [reply src/sentry/middleware/proxy.py:110]: Ok, this has been resolved.

**Review by wedamija — APPROVED** [review]

**Pull request merged**